remme/tp/node_account.py

```diff
@@ -0,0 +1,103 @@
+        return {
+            NodeAccountMethod.INITIALIZE_MASTERNODE: {
+                PB_CLASS: NodeAccountInternalTransferPayload,
+                PROCESSOR: self._initialize_masternode,
+                VALIDATOR: NodeAccountInternalTransferPayloadForm,
+            },
+            NodeAccountMethod.CLOSE_MASTERNODE: {
+                PROCESSOR: self._close_masternode,
+                PB_CLASS: EmptyPayload,
+                VALIDATOR: EmptyPayloadForm,
+            },
+            NodeAccountMethod.TRANSFER_FROM_UNFROZEN_TO_OPERATIONAL: {
+                PB_CLASS: NodeAccountInternalTransferPayload,
+                PROCESSOR: self._transfer_from_unfrozen_to_operational,
+                VALIDATOR: NodeAccountInternalTransferPayloadForm,
+            }
+        }
+
+    def _initialize_masternode(self, context, node_account_public_key, internal_transfer_payload):
+        node_account_address = self.make_address_from_data(node_account_public_key)
+
+        node_account = get_data(context, NodeAccount, node_account_address)
+
+        if node_account.node_state != NodeAccount.NodeState.Value('NEW'):
+            raise InvalidTransaction('Masternode is already opened or closed.')
+        minimum_stake = int(_get_setting_value(context, SETTINGS_MINIMUM_STAKE))
+
+        if internal_transfer_payload.value < minimum_stake:
+            raise InvalidTransaction('Initial stake is too low.')
+
+        node_account.node_state = NodeAccount.NodeState.Value('OPENED')
+
+        node_account.balance -= internal_transfer_payload.value
+
+        unfrozen_part = internal_transfer_payload.value - minimum_stake
+        node_account.reputation.frozen += minimum_stake
+        node_account.reputation.unfrozen += unfrozen_part
+
+        return {
+            node_account_address: node_account,
+        }
+
+    def _close_masternode(self, context, node_account_public_key, payload):
+        node_account_address = self.make_address_from_data(node_account_public_key)
+
+        node_account = get_data(context, NodeAccount, node_account_address)
+
+        if node_account.node_state != NodeAccount.NodeState.Value('OPENED'):
+            raise InvalidTransaction('Masternode is not opened or has been closed.')
+
+        node_account.node_state = NodeAccount.NodeState.Value('CLOSED')
+
+        node_account.balance += node_account.reputation.frozen
+        node_account.balance += node_account.reputation.unfrozen
+
+        node_account.reputation.frozen = 0
+        node_account.reputation.unfrozen = 0
+
+        return {
+            node_account_address: node_account,
+        }
+
+    def _transfer_from_unfrozen_to_operational(self, context, node_account_public_key, internal_transfer_payload):
+
+        node_account_address = self.make_address_from_data(node_account_public_key)
+
+        node_account = get_data(context, NodeAccount, node_account_address)
+
+        node_account.reputation.unfrozen -= internal_transfer_payload.value
```

Comment:
Check if `node_account.reputation.unfrozen >= internal_transfer_payload.value`

Reply:
I had test for that and sawtooth throws ValueError exception because this field is uint. But I agree that it will provide more security so I added this condition into the code.